lil/src/Interpreter.hs

```
@@ -0,0 +1,105 @@
+        then throwError "Divide by zero"
+        else return $ lhs' `mod` rhs'
+    Eq -> return $ fromEnum $ lhs' == rhs'
+    Ne -> return $ fromEnum $ lhs' /= rhs'
+    Gt -> return $ fromEnum $ lhs' > rhs'
+    Ge -> return $ fromEnum $ lhs' >= rhs'
+    Lt -> return $ fromEnum $ lhs' < rhs'
+    Le -> return $ fromEnum $ lhs' <= rhs'
+    And -> return $ fromEnum $ lhs' /= 0 && rhs' /= 0
+    Or -> return $ fromEnum $ lhs' /= 0 || rhs' /= 0
+
+evalExpr :: Expr -> EvalResult
+evalExpr expr =
+  case expr of
+    Const n -> return n
+    Var x -> do
+      vars' <- gets vars
+      let result = lookup x vars'
+      case result of
+        Just n -> return n
+        Nothing -> throwError $ "Undefined variable: " ++ x
+    Call f args -> do
+      funcs' <- gets funcs
+      let result = find (\(Func name _ _) -> name == f) funcs'
+      case result of
+        Just func@(Func _ params _) ->
+          if length params /= length args
+            then throwError $ "Insufficient number of arguments passed to function " ++ f
+            else callFunc func args
+        Nothing -> throwError $ "Undefined function: " ++ f
+    Binop op lhs rhs -> evalBinop op lhs rhs
+  where
+    callFunc (Func _ params body) args = do
+      args' <- mapM evalExpr args
+      let newVars = zip params args'
+      vars' <- gets vars
+      modify (\env -> env {vars = newVars})
+      value <- evalStmt body
+      modify (\env -> env {vars = vars'})
+      return value
+
+evalStmt :: Stmt -> EvalResult
+evalStmt stmt =
+  case stmt of
+    Asgn name expr -> do
+      value <- evalExpr expr
+      modify (\env -> env {vars = (name, value) : vars env})
+      return value
+    Expr' expr -> evalExpr expr
+    Write expr -> do
+      value <- evalExpr expr
+      handle' <- gets handle
+      liftIO $ hPrint handle' value
+      return value
+    Read name -> do
+      value <- liftIO getLine
+      modify (\env -> env {vars = (name, read value) : vars env})
+      return 0
```

Here it's better to report an error, if a person inputs anything but a number.

---

And maybe ask them to try again